Generator: Use api.NewURL for URL generation

masnax opened the pull request:
Cleaning up URLs from #9364, I've made some changes to consolidate URL generation to always use the methods in `shared/api/url.go`.

Main changes:
* Remove generation of prepared statements for getting names and projects of entities, preferring instead to just use the pre-existing prepared statements for getting whole entities, but just returning name and project via go.
* Omit `target=none` from URLs with `target` form values.
* Move `urlsToResourceNames` to `shared/api/url.go` to have all URL conversion methods in one place.
* Remove db package `EntityFormatURIs` and `SelectURIs` in favour of using `api.NewURL` to create URLs.

stgraber commented:
Seems to be breaking `lxc project list` at least.

masnax commented:
> Seems to be breaking `lxc project list` at least.

Was missing a generator comment for getting projects by ID. At least generator issues don't fail silently now :)

stgraber commented:
jenkins: test this please

tomponline commented:
@masnax in the commit " lxd/db/generate/db/method: Generate URIs from get-objects stmt " - why is the DB generator still needing to generate URLs at all? I thought we were moving the UserBy population out of the DB generator and AFAIK that is the only thing that needs to generate URLs?

tomponline commented on shared/api/url.go:
```diff
@@ -57,7 +57,7 @@ func (u *URL) Project(projectName string) *URL {
 // Target sets the "target" query parameter in the URL if the clusterMemberName is not empty or "default".
 func (u *URL) Target(clusterMemberName string) *URL {
-	if clusterMemberName != "" {
+	if clusterMemberName != "" && clusterMemberName != "none" {
```
Please can you add a comment as to how this can come about? Where does "none" come from?

masnax replied:
So prior to the changes to the generator that were just pushed, when `UsedBy` fields were populated for Projects, we were stripping out `target=none` and `project=default`. I just moved it over assuming it was important:

```go
	for j := range value {
			if len(value[j]) > 12 && value[j][len(value[j])-12:] == "&target=none" {
				value[j] = value[j][0 : len(value[j])-12]
			}
			if len(value[j]) > 16 && value[j][len(value[j])-16:] == "?project=default" {
				value[j] = value[j][0 : len(value[j])-16]
			}
		}
```

You can see the diff here: https://github.com/lxc/lxd/commit/13def65775a19f38fd268ad720605a97e3c3d282#diff-5b408e12c31e2c3614dce882c6e52920de957b2565c76bbd86b06a940eea1d0a 
under `lxd/db/projects.mapper.go` line 161.

I think it's for storage pools/volumes, because we're filtering out `nodesName[*r.nodeID] == "none"` in `lxd/db/storage_pools.go`:
https://github.com/lxc/lxd/blob/master/lxd/db/storage_pools.go#L138-L153

tomponline replied:
Hrm OK thats a new one to me.

@stgraber any idea where `target=none` is used? Is that still relevant?

tomponline commented:
@masnax also in " lxd/db/generate/db/method: Generate URIs from get-objects stmt " whilst I'm in principle all for using object load function for URL generation rather than needing to a dedicated name loading function, do the get object functions also load config? Or to put it another way, do the get-objects functions only load the rows from the primary object table, or do they also perform joins or additional queries on other tables that may make switching to these types of function more inefficient.

tomponline commented on shared/api/url.go:
```diff
@@ -65,3 +66,25 @@ func (u *URL) Target(clusterMemberName string) *URL {
+
+// UrlsToResourceNames returns a list of resource names extracted from one or more URLs of the same resource type.
+// The resource type path prefix to match is provided by the matchPathPrefix argument.
+func UrlsToResourceNames(matchPathPrefix string, urls ...string) ([]string, error) {
```
now we are exporting this function, it should be `URLsToResourceNames` so that acronyms are capitalised.

https://talks.golang.org/2014/names.slide#5

masnax replied:
Done. PR should be ready to merge now.

masnax commented:
> @masnax in the commit " lxd/db/generate/db/method: Generate URIs from get-objects stmt " - why is the DB generator still needing to generate URLs at all? I thought we were moving the UserBy population out of the DB generator and AFAIK that is the only thing that needs to generate URLs?

Guess I'm still in "minimize-the-diff" mode when it comes to the generator. I tried to not refactor too much of the logic, rather just swap out the methods that were being used. 

Alternatively I can convert the `GetURIs` methods to `GetNameAndProject`, and convert the result to a URL wherever we want to use one.

> @masnax also in " lxd/db/generate/db/method: Generate URIs from get-objects stmt " whilst I'm in principle all for using object load function for URL generation rather than needing to a dedicated name loading function, do the get object functions also load config? Or to put it another way, do the get-objects functions only load the rows from the primary object table, or do they also perform joins or additional queries on other tables that may make switching to these types of function more inefficient.

Actually I'm not reusing the get-objects _function_ to generate URLs, I'm reusing the get-objects _statement_ in a new, minimal function. The actual SQL statement makes no mention of the reference fields so config/devices/etc won't be filled. Ultimately the amount of queries doesn't change.

tomponline commented:
@masnax I can understand that desire. :)

So if this is a stepping stone that's fine, I would like to understand where we are aiming to get to though.

Broadly speaking, I can see 4 main access patterns:

1. Load a single object (just its direct table fields, without config and used by).
2. Load a list of objects (just its direct table fields, without config and used by).
3. Load a single object and its info (direct table fields, with config and used by)
4. Load a list of objects and their info (direct table fields, with config and used by)

Note, I've been specific not to mention URLs anyway in those access patterns, as URLs seems to be to be outside of the scope of the DB layer, and something that falls within the so-called presentation/view layer (they can be generated from object type, name of the object and project of the object). 

Whether we are generating a structured table output, CSV, YAML, JSON or URL, these are just different presentation formats, that all use the same base data from the database. We don't have the DB layer generating JSON or YAML (quite rightly), but for some historical reason we do have the DB layer generating URLs, and this seems to be the odd one out.

In an ideal world we wouldn't need 4 access patterns, and could just have 3 and 4 (fully loading the objects and their config and used by even when we don't need the data). But its inefficient and can slow down queries on larger data sets. This becomes even more of a problem with dqlite where the cluster member running the query may be remote to the leader database and thus each transaction and query must incur network round-trip costs. Thus we do not want to be unnecessarily load config and used by data if we are not using it (because we are producing a URL list of resources for example).

So I propose that we aim (possibly in several steps) to remove all URL generation from the DB layer.

But we still need functions that allow returning of just the primary object fields (without their config and used by data), so that the URLs can be generated from the data without additional overhead.

I suggest 2 possible approaches, and for this I will use Profiles as the example object type:

Approach 1:
Create a `Profile` struct that just contains `Project` and `Name` (and possibly ID) fields.
Then create a `ProfileInfo` struct that embeds the `Profile` struct, along with all the other fields from the profiles table, along with the config.

We would then have functions that return either `Profile` or `ProfileInfo` structs (in either singular or list form). 
And we would use the former to generate URLs from (we could even at that point at a `URL` function to the `Profile` type to make it super easy to convert an object into a URL).

Approach 2:

We would leave the `Profile` struct as it is today, and for the function that just loads the project and name of an object we would return it using native structures, such as `map[projectName][]profileName`.

In both approaches, the UsedBy field population of either `ProfileInfo` (approach 1) or `Profile` (approach 2) structs would be moved out of the DB generator to a manually written function.

Is that broadly inline with what you were planning?

masnax commented:
> I suggest 2 possible approaches, and for this I will use Profiles as the example object type:
> 
> Approach 1: Create a `Profile` struct that just contains `Project` and `Name` (and possibly ID) fields. Then create a `ProfileInfo` struct that embeds the `Profile` struct, along with all the other fields from the profiles table, along with the config.
> 
> We would then have functions that return either `Profile` or `ProfileInfo` structs (in either singular or list form). And we would use the former to generate URLs from (we could even at that point at a `URL` function to the `Profile` type to make it super easy to convert an object into a URL).
> 
> Approach 2:
> 
> We would leave the `Profile` struct as it is today, and for the function that just loads the project and name of an object we would return it using native structures, such as `map[projectName][]profileName`.
> 
> In both approaches, the UsedBy field population of either `ProfileInfo` (approach 1) or `Profile` (approach 2) structs would be moved out of the DB generator to a manually written function.
> 
> Is that broadly inline with what you were planning?

Right, we're on the same page :)

I'd also like suggest an Approach 3:

Strip out reference fields from `Profiles` entirely so that `GetProfile` always queries exactly one table. Then add reference methods like:
```go
GetProfileDevices(p Profile) (map[string]Device, error)
GetProfileConfig(p Profile) (map[string]string, error)
GetProfileUsedBy(p Profile) ([]string, error)
```

Thus we can get each bit of extra info only when we need it. 

I'll have to see which one of these three versions will make the most sense based on how we use reference fields today/

tomponline commented:
OK good.

So we could have a: `(p Profile) URL() net.URL`

Some notes of caution/note though:

1. I difference between the `URL()` function and the `GetDevices` and `GetConfig` functions is that the latter two need to run additional queries. So we don't want to be in a position that these functions start a new transaction each time they are called, as if they are called as part of a list iterator this will cause many transactions to be started, which are expensive. So we need to have a way to start a single transaction, retrieve initial object list, and then (if needed) iterate through the list getting the config and or device info, but within the same transaction.
2. I don't think we should have a `UsedBy` field in the DB structs at all, this should be left in the api structs for the objects, and then populated using a manual (non-generated) function when needed after retrieving the objects from the DB layer. This way the DB generator does not need to have the complexity of understanding relationships between objects, and instead can focus on reducing code duplication for the basic CRUD queries.

masnax commented:
>     1. I difference between the `URL()` function and the `GetDevices` and `GetConfig` functions is that the latter two need to run additional queries. So we don't want to be in a position that these functions start a new transaction each time they are called, as if they are called as part of a list iterator this will cause many transactions to be started, which are expensive. So we need to have a way to start a single transaction, retrieve initial object list, and then (if needed) iterate through the list getting the config and or device info, but within the same transaction.

Hm, I guess the only solutions there would be to:
* Make the methods belong to a `ClusterTx` transaction.
* Pass a `tx *sql.Tx` as an argument to those methods.
* Have a helper that takes a slice `[]Profile`.

I'll play around with it to see which makes more sense.

As for the URL method. In keeping with the Profile example, do you think it would make more sense to have the URL method be on `api.Profile` instead of `db.Profile`? This way, if we ever want to generate the URL, we would then first have to convert to the api struct first, which we already have methods for.

>     2. I don't think we should have a `UsedBy` field in the DB structs at all, this should be left in the api structs for the objects, and then populated using a manual (non-generated) function when needed after retrieving the objects from the DB layer. This way the DB generator does not need to have the complexity of understanding relationships between objects, and instead can focus on reducing code duplication for the basic CRUD queries.

Right, I would go even further and suggest to remove all the fields that reference other tables on any of the LXD entity structs in the db package, and convert them all to methods like `profile.UsedBy()`, `profile.Devices()`, `instance.Profiles()`, `instance.Config()`. I think this is beyond the scope of this PR though, I'm starting a branch for all this.

tomponline commented:
> As for the URL method. In keeping with the Profile example, do you think it would make more sense to have the URL method be on `api.Profile` instead of `db.Profile`? This way, if we ever want to generate the URL, we would then first have to convert to the api struct first, which we already have methods for.

> EDIT: Actually, I think adding `URL()` to the `api` package structs might make things a bit messy because not every database entity maps cleanly to its api equivalent, namely instances.

I think adding a `URL()` function to the structs in the `api` package is a good idea. This would be helpful to clients as well as LXD server itself, and would reinforce that the URL is a "front-end" presentation concept and not a DB concept. Good idea.

I'm not clear what your EDIT meant as to why you think this wouldn't work though?

masnax commented:
> I'm not clear what your EDIT meant as to why you think this wouldn't work though?

It would work, just be a bit more involved I think, as how we access URLs would affect where we put the UsedBy() methods. Those methods would probably have to leave the db package as well in this case. Currently this is how it is for networks, which you've worked on already (the UsedBy methods are under `lxd/networks`).

tomponline commented:
> It would work, just be a bit more involved I think, as how we access URLs would affect where we put the UsedBy() methods. Those methods would probably have to leave the db package as well in this case. Currently this is how it is for networks, which you've worked on already (the UsedBy methods are under `lxd/networks`).

Yes I don't have a problem with that per-se, as generating UsedBy info is not a pure DB function, they will use multiple DB function calls internally to get a list of various resources, and then inspect them (sometimes by decoding JSON held within the object) for whether they are using the resource in question. 

One issue I can see with adding `URL()` to the API object struct is that they don't have a concept of project in them.
However this doesn't seem much of a problem as you could have the `URL()` function return an `api.URL` which could then be chained as so:

```go
apiProfile.URL().Project("foo").String()
```

As whenever you're dealing with an API object type you always have the project to-hand already.

tomponline commented:
@masnax I thought a real-life example might help out thinking, and doing so has made me come round to the idea of a `URL` function on the DB struct type (and means at least for now we wouldn't need one on the API struct type).

What do you think of this:

```go
func profilesGet(d *Daemon, r *http.Request) response.Response {
	projectName, _, err := project.ProfileProject(d.State().Cluster, projectParam(r))
	if err != nil {
		return response.SmartError(err)
	}

	recursion := util.IsRecursionRequest(r)

	var result interface{}
	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
		filter := db.ProfileFilter{
			Project: &projectName,
		}

		profiles, err := tx.GetProfiles(filter)
		if err != nil {
			return err
		}

		if recursion {
			apiProfiles := make([]*api.Profile, len(profiles))
			for i, profile := range profiles {
				apiProfiles[i] = profile.API(tx)                   // Will then call profile.Devices(tx) and profile.Config(tx) internally.
				apiProfiles[i].UsedBy = profileUsedBy(tx, profile) // Manual function.
			}

			result = apiProfiles
		} else {
			apiURLs := make([]string, len(profiles))

			for i, profile := range profiles {
				apiURLs[i] = profile.URL() // Will then use api.NewURL internally.
			}

			result = apiURLs
		}
		return err
	})
	if err != nil {
		return response.SmartError(err)
	}

	return response.SyncResponse(true, result)
}
```

It achieves the following:

- All queries run in a single transaction (with the possibility that the queries themselves are using prepared statements).
- The population of the API `UsedBy` field is manual and outside of the DB package.
- We don't load the config and devices for URL generation.
- There are no specific queries for URL generation.

masnax commented:
> @masnax any thoughts on [#9618 (comment)](https://github.com/lxc/lxd/pull/9618#issuecomment-985388273)? I chatted with @stgraber about it the other day and he said it was fine. If you don't see any problems with it, we can work out the steps to achieve it.

I’ve already been working through some steps to achieve it :)

I have a branch going where I’ve moved all of the extra-table fields out of the db package structs, and converted them to helper methods that take a tx cluster transaction.

One quirk is that db.Instance doesn’t map cleanly to api.Instance. This is mainly due to snapshots and instance state. Right now I’m using a new struct db.InstanceFull to store the extra fields we want (devices, config, profiles). Everywhere else though, the api package struct can be used whenever we want all extra-table fields.

I’ve also tried to move URLs to the api package. In most cases this means that the version.APIVersion constant needs to be passed in as an argument as we would have to import that package into shared/api otherwise.  

Also, as a side effect, this has made cleaning up the previous changes to expanded devices (with a lot of redundant type conversions) very easy!

tomponline commented:
@masnax excellent!

Can you clarify with an example what you mean by that? 
> One quirk is that db.Instance doesn’t map cleanly to api.Instance. This is mainly due to snapshots and instance state.

tomponline commented:
@masnax also are you happy to merge this PR as it currently is and then send follow up PRs?

masnax commented:
> @masnax also are you happy to merge this PR as it currently is and then send follow up PRs?

I have some small changes I'd like to push first, but yes, I think this PR has accomplished removing the old URL generation so it can be merged.

tomponline commented:
thanks if you can fix https://github.com/lxc/lxd/pull/9618/files#r761516550 then im happy to merge.

masnax commented:
> @masnax excellent!
> 
> > One quirk is that db.Instance doesn’t map cleanly to api.Instance. This is mainly due to snapshots and instance state.
> 
> Can you clarify with an example what you mean by that?

`api.Instance` has the fields `InstancePut.Restore`, `Status` and 'StatusCode'. They can be filled when converting from `db.Instance.ToAPI()`, but for the purposes we're going for here (accessing config, profiles, devices), they would never be used. I also didn't want to leave them empty since that would mean that the returned `api.Instance` is incomplete.

tomponline commented:
@masnax some conflicts to resolve then good to merge.

masnax commented:
> @masnax some conflicts to resolve then good to merge.

Should be good to go now.

tomponline commented on client/lxd_certificates.go:
```diff
@@ -20,7 +20,7 @@ func (r *ProtocolLXD) GetCertificateFingerprints() ([]string, error) {
 
 	// Parse it.
-	return urlsToResourceNames(baseURL, urls...)
+	return api.URLsToResourceNames(baseURL, urls...)
```
@masnax given what we've discussed now, is it possible to remove this commit, as I don't want to move it to the `share/api` package only to remove it shortly after (as it is considered a public package that others may depend on).

masnax replied:
Sure. I'd have to duplicate (or make public, but that would have a larger diff) `urlsToResourceNames` in the db package temporarily.

tomponline replied:
Yeah client is shared too, so making it public is no good. Happy for it to be duplicated if only for a short time.

masnax replied:
Tests seem to be passing so I think it's good to merge now.
